cloudinit/sources/DataSourceOracle.py

```diff
@@ -292,72 +276,14 @@ def _is_iscsi_root():
     return bool(cmdline.read_initramfs_config())
 
 
-def _load_index(content):
-    """Return a list entries parsed from content.
-
-    OpenStack's metadata service returns a newline delimited list
-    of items.  Oracle's implementation has html formatted list of links.
-    The parser here just grabs targets from <a href="target">
-    and throws away "../".
-
-    Oracle has accepted that to be buggy and may fix in the future
-    to instead return a '\n' delimited plain text list.  This function
-    will continue to work if that change is made."""
-    if not content.lower().startswith("<html>"):
-        return content.splitlines()
-    items = re.findall(
-        r'href="(?P<target>[^"]*)"', content, re.MULTILINE | re.IGNORECASE)
-    return [i for i in items if not i.startswith(".")]
-
-
-def read_metadata(endpoint_base=METADATA_ENDPOINT, sys_uuid=None,
-                  version='2013-10-17'):
-    """Read metadata, return a dictionary.
-
-    Each path listed in the index will be represented in the dictionary.
-    If the path ends in .json, then the content will be decoded and
-    populated into the dictionary.
-
-    The system uuid (/sys/class/dmi/id/product_uuid) is also populated.
-    Example: given paths = ('user_data', 'meta_data.json')
-    This would return:
-      {version: {'user_data': b'blob', 'meta_data': json.loads(blob.decode())
-                 'system_uuid': '3b54f2e0-3ab2-458d-b770-af9926eee3b2'}}
+def read_opc_metadata():
     """
-    endpoint = combine_url(endpoint_base, version) + "/"
-    if sys_uuid is None:
-        sys_uuid = _read_system_uuid()
-    if not sys_uuid:
-        raise sources.BrokenMetadata("Failed to read system uuid.")
-
-    try:
-        resp = readurl(endpoint)
-        if not resp.ok():
-            raise sources.BrokenMetadata(
-                "Bad response from %s: %s" % (endpoint, resp.code))
-    except UrlError as e:
-        raise sources.BrokenMetadata(
-            "Failed to read index at %s: %s" % (endpoint, e))
-
-    entries = _load_index(resp.contents.decode('utf-8'))
-    LOG.debug("index url %s contained: %s", endpoint, entries)
-
-    # meta_data.json is required.
-    mdj = 'meta_data.json'
-    if mdj not in entries:
-        raise sources.BrokenMetadata(
-            "Required field '%s' missing in index at %s" % (mdj, endpoint))
-
-    ret = {'system_uuid': sys_uuid}
-    for path in entries:
-        response = readurl(combine_url(endpoint, path))
-        if path.endswith(".json"):
-            ret[path.rpartition(".")[0]] = (
-                json.loads(response.contents.decode('utf-8')))
-        else:
-            ret[path] = response.contents
+    Fetch metadata from the /opc/ routes.
 
-    return {version: ret}
+    :return:
+        The JSON-decoded value of the /opc/v1/instance/ endpoint on the IMDS.
+    """
+    return json.loads(readurl(METADATA_ENDPOINT, retries=2)._response.text)
```

Comment:
2? just because 2 is better than one?

I think we have a pretty convincing argument that we're running on OracleCloud at this point ('OracleCloud.com') any reason to not bump that up? Or turn it to no retries.  2 just seems an odd number here.  Either we're dependent on it (no retries) or we want to be resilient (10 or more?).  

I don't have strong feelings on this though.

Reply:
i knew that [two is an impossible number](http://wiki.c2.com/?TwoIsAnImpossibleNumber), but this interpretation is new to me ;)

Reply:
`retries=2` was requested by Oracle in a code review to handle a potential race condition where the IMDS might not be ready. Re-reading their point I think they really wanted 2 _tries_ not 2 retries (and the race condition window is reportedly milliseconds, so a single retry will comfortably address it). I'll modify this to 1.

Reply:
hmm... well thanks for pointing out that it is *re*tries, rather than tries. The same argument applies though, just with N-1.

> handle a potential race condition where the IMDS might not be ready. 

In other places (EC2) we have similar behavior.  We `wait_for_metadata_service` only on the first request, and then subsequent calls the IMDS must be there, and readurl gets the default `retries=0`.

> and the race condition window is reportedly milliseconds, so a single retry will comfortably address it

"The IMDS is not yet ready" seems unlikely to reliably fix itself milliseconds later.  The cloud platform prepared gigabytes of disk, booted a virtual (or physical machine), loaded a kernel, initramfs, switchroot, init, dhcp, .... all of this stuff happened, but the system beat the IMDS up by *milliseconds* ?  Just seems unlikely.

I wont fight that, please at least document it in code comment though.